File: backend/app/schemas/patient.py
from datetime import date
from uuid import UUID

from pydantic import BaseModel


class PatientSummaryResponse(BaseModel):
    id: UUID
    pid: str | None = None
    citizen_id: str | None = None
    hn: str | None = None
    full_name: str
    birth_date: date | None = None


class DiagnosisHistoryResponse(BaseModel):
    visit_date: date
    diagnosis_code: str | None = None
    diagnosis_name: str | None = None
    normalized_disease_key: str | None = None
    department: str | None = None
    doctor_name: str | None = None


class PatientHistoryResponse(BaseModel):
    patient: PatientSummaryResponse
    history: list[DiagnosisHistoryResponse]


# Phase C: per-record detail from disease_screening_records (new import pipeline table)
class ScreeningRecordResponse(BaseModel):
    record_id: UUID
    source_file_name: str | None = None
    source_row_no: int | None = None
    normalized_person_identifier: str
    full_name: str | None = None
    raw_service_type: str
    normalized_service_key: str
    visit_date: date


# Phase C: combined source-history response for a target group result row.
# Contains screening DB records + TG file history events in separate buckets
# so the UI can render them in clearly labelled sections.
class ResultSourceHistoryResponse(BaseModel):
    result_id: UUID
    normalized_cid: str | None = None
    full_name: str | None = None
    screening_db_records: list[ScreeningRecordResponse]
    target_group_history_events: list[dict]
    history_source_summary: str

File: frontend/src/types/patient.ts
export type PatientSummary = {
  id: string;
  pid: string | null;
  citizen_id: string | null;
  hn: string | null;
  full_name: string;
  birth_date: string | null;
};

export type DiagnosisHistoryRow = {
  visit_date: string;
  diagnosis_code: string | null;
  diagnosis_name: string | null;
  normalized_disease_key: string | null;
  department: string | null;
  doctor_name: string | null;
};

export type PatientHistory = {
  patient: PatientSummary;
  history: DiagnosisHistoryRow[];
};

// Phase C / source-history endpoint types ----------------------------------

/** One row from disease_screening_records, returned by source-history endpoint. */
export type ScreeningRecord = {
  record_id: string;
  source_file_name: string | null;
  source_row_no: number | null;
  normalized_person_identifier: string;
  full_name: string | null;
  raw_service_type: string;
  normalized_service_key: string;
  visit_date: string;
};

/** One event row from target_group_history_rows, returned by source-history endpoint. */
export type TgHistoryEvent = {
  source_type: string | null;
  source_file_name: string | null;
  source_sheet_name: string | null;
  source_row_no: number | null;
  raw_service_type: string | null;
  normalized_service_key: string | null;
  visit_date: string | null;
  raw_result: string | null;
  raw_hospital: string | null;
  raw_doctor: string | null;
  raw_note: string | null;
  validation_status: string | null;
  warning_message: string | null;
};

/**
 * Response from GET /api/target-groups/{group_id}/results/{result_id}/source-history.
 * Contains BOTH evidence buckets for one result row so the UI can display them
 * in clearly labelled sections without any de-duplication confusion.
 */
export type ResultSourceHistory = {
  result_id: string;
  normalized_cid: string | null;
  full_name: string | null;
  screening_db_records: ScreeningRecord[];
  target_group_history_events: TgHistoryEvent[];
  history_source_summary:
    | "both_sources"
    | "screening_db_only"
    | "target_group_file_only"
    | "no_history_found";
};
